firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,446 @@
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+      THEN("The second transform method reports ok status")
+      THEN("After the second transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+/* Datagram sender
+
+Scenario: Datagram Sender correctly generates datagram bodies from payloads
+  GIVEN("A Datagram sender of buffer size 254 bytes with next sequence equal to 0")
+    WHEN("A non empty payload buffer is given to the datagram sender")
+      THEN("The transform method reports ok status")
+      THEN("The seq field of the body's header is equal to the next sequence")
+      THEN("The length field of the body's header is equal to the size of the payload buffer given")
+      THEN("The body's payload section correctly stores the paylaod as '0xf9 0x23 0x4a 0xd4 0xe0'")
+      THEN("The output datagram is as expected '0x00 0x05 0xf9 0x23 0x4a 0xd4 0xe0'")
+
+    WHEN("The output datagram cannot hold enough data")
+      THEN("The transform method reports to invalid length")
+      THEN("The output buffer is as expected")
+
+    WHEN("The input paylaod is too large for the output datagram to hold")
+      THEN("The transform method reports invalid length")
+      THEN("The output buffer is as expected")
+    
+    WHEN("The input payload to the datagram sender is '0x00 0x00'")
+      THEN("The transform method reports ok status")
+      THEN("The output datagram is as expected '\x00\x02\x00\x00' ")      
+
+Scenario: Datagram Sender correctly reports incrementing and rollover of expected sequence number
+  GIVEN("A Datagram sender of buffer size 254 bytes with next sequence equal to 0")
+    
+    WHEN("2 non-empty payload buffers of capacity 254 bytes is given to the sender")
+      THEN("The first transform method reports ok status")      
+      THEN("The seq field of the body's header is equal to the next sequence")
+      THEN("The length field of the body's header is equal to the size of the payload given")        
+      THEN("The body's payload section correctly stores the paylaod as '0xc0 0x18 0x65 0xd1 0x03 0x5c'")
+      THEN("The output datagram is as expected '0x00 0x06 0xc0 0x18 0x65 0xd1 0x03 0x5c' ")
+
+      THEN("The second transform method reports ok status")      
+      THEN("The seq field of the body's header is equal to the next sequence")
```

Comment:
GIVEN defines the next sequence as 0. This "next sequence" is different from the "next sequence" on line 422, so we should specify what it actually is.

Reply:
I'd suggest having a test for seq 1 to seq 2. Right now we have seq 0 to seq 1, and seq 255 to seq 0; both of them are edge cases since they involve seq 0. seq 1 to seq 2 would be kind of like the inductive step in an inductive proof.

Reply:
Fixed.